rule CGR:
    input:
        "../data/genomes/{species}_genomes.fna"
    output:
        "../data/following/CGRs/{windows}/{species}_done.txt"
    threads:
        99
    shell:
        "python3 ../scripts/windowed_CGR.py {input} {wildcards.windows} \
            {threads} {output}"

rule CDS:
    input:
        "../data/genomes/{species}_genomes.fna",
        "../data/features/genes/{species}_feature_table.txt",
    output:
        "../files/features/{windows}/{species}_CDS.txt"
    shell:
        "python3 ../scripts/extract_CDS.py {input} {wildcards.windows} \
            {output}"

rule LCR:
    input:
        "../data/genomes/{species}_genomes.fna",
        "../data/features/repeats/{species}_repeats.txt",
    output:
        "../files/features/{windows}/{species}_LCR.txt"
    shell:
        "python3 ../scripts/extract_LCR.py {input} {wildcards.windows} \
            {output}"

rule RR:
    input:
        "../data/genomes/{species}_genomes.fna",
        "../data/features/recombination_rates/{species}_RR_spline.RData",
    output:
        "../files/features/{windows}/{species}_RR.txt"
    shell:
        "python3 ../scripts/extract_RR.py {input} {wildcards.windows} \
            {output}"

rule ratio_extract:
    input:
        "../data/following/CGRs/{windows}/{species}_done.txt"
    output:
        "../files/features/{windows}_ratios/{species}.txt"
    shell:
        "python3 ../scripts/nucleotide_ratio_extract.py {input} \
            {wildcards.species} {wildcards.windows} {threads} {output}"

rule FCGR:
    input:
        "../data/following/CGRs/{windows}/{species}_done.txt"
    output:
        "files/FCGRs/{windows}_{kmer}/{species}_FCGRs.txt"
    threads:
        99
    shell:
        "python3 scripts/windowed_FCGR.py {input} {wildcards.species} \
            {wildcards.windows} {wildcards.kmer} {threads} {output}"

rule FCGR_dist:
    input:
        "files/FCGRs/{windows}_{kmer}/{species}_FCGRs.txt"
    output:
        "files/distances/pearson/{windows}_{kmer}/{species}_dist_matrix.RData"
    threads:
        99
    shell:
        "Rscript scripts/pearson_distance.R {input} {output}"

rule correlation:
    input:
        "files/distances/pearson/{windows}_{kmer}/{species}_dist_matrix.RData",
        "../files/features/{windows}/{species}_{features}.txt"
    output:
        "files/results/{windows}_{kmer}_{features}/{species}_correlation.png"
    shell:
        "Rscript ../scripts/cor_dist_feature.R {output} {input} FCGRs \
            {wildcards.features} {wildcards.windows} {wildcards.kmer}"

rule MDS:
    input:
        "files/distances/pearson/{windows}_{kmer}/{species}_dist_matrix.RData",
        "../files/features/{windows}/{species}_{features}.txt"
    output:
        "files/results/{windows}_{kmer}_{features}/{species}_MDS.png"
    shell:
        "Rscript ../scripts/MDS_dist_feature.R {output} {input} FCGRs \
            {wildcards.features} {wildcards.windows} {wildcards.kmer}"

rule ratio_figure:
    input:
        "files/distances/pearson/{windows}_{kmer}/{species}_dist_matrix.RData",
        "../files/features/{windows}_ratios/{species}.txt"
    output:
        "files/results/{windows}_{kmer}_ratios/{species}.png"
    shell:
        "Rscript ../scripts/nucleotide_ratio_figures.R {output} {input} \
            {wildcards.windows} FCGRs {wildcards.kmer}"
