################################################################################
### Part 0.1) Downloading part
################################################################################

rule download_genomes:
    output:
        "../data/genomes/{species}_genomes.fna"
    shell:
        "bash ../scripts/download_genomes.sh {wildcards.species} {output}"

rule download_features:
    output:
        "../data/factors/features/{species}_feature_table.txt"
    shell:
        "bash ../scripts/download_features.sh {wildcards.species} {output}"

rule download_genes:
    output:
        "../data/factors/genes/{species}_genes.gff"
    shell:
        "bash ../scripts/download_genes.sh {wildcards.species} {output}"

rule download_repeats:
    output:
        "../data/factors/repeats/{species}_repeats.txt"
    shell:
        "bash ../scripts/download_repeats.sh {wildcards.species} {output}"

rule download_rr:
    output:
        "../data/factors/recombination_rates/{species}_RR_spline.RData"
    shell:
        "bash ../scripts/download_repeats.sh {wildcards.species} {output}"

################################################################################
### Part 0.2) Extract and clean ranges
################################################################################

rule factor_proxies:
    input:
        "../data/genomes/{species}_genomes.fna",
        "../data/factors/repeats/{species}_repeats.txt",
        "../data/factors/features/{species}_feature_table.txt",
        "../data/factors/genes/{species}_genes.gff"
    output:
        "../data/following/factor_proxies/{species}/{factors}_proxies_done.txt"
    shell:
        "python3 ../scripts/factor_proxies.py {wildcards.factors} {input} \
        {output}"

rule extract_overlaps:
    input:
        "../data/genomes/{species}_genomes.fna"
    output:
        "../data/following/factor_proxies/overlaps/{species}_done.txt"
    threads:
        99
    shell:
        "python3 ../scripts/extract_overlaps.py {input} {threads} {output}"

rule clean_ranges:
    input:
        "../data/genomes/{species}_genomes.fna",
        "../data/following/factor_proxies/overlaps/{species}_done.txt"
    output:
        "../data/following/factor_proxies/cleaning/{species}_done.txt"
    threads:
        99
    shell:
        "python3 ../scripts/removing_overlaps.py {input} {threads} {output}"

rule extract_uncategorized:
    input:
        "../data/genomes/{species}_genomes.fna",
        "../data/following/factor_proxies/cleaning/{species}_done.txt"
    output:
        "../data/following/factor_proxies/uncategorized/{species}_done.txt"
    threads:
        99
    shell:
        "python3 ../scripts/extract_uncategorized.py {input} {threads} {output}"

rule factor_percentages:
    input:
        "../data/genomes/{species}_genomes.fna",
        "../data/following/factor_proxies/uncategorized/{species}_done.txt"
    output:
        "../files/factor_percentages/{windows}/{species}.txt"
    threads:
        99
    shell:
        "python3 ../scripts/factor_percentages.py {input} {wildcards.windows} \
        {output}"

rule filter_factors:
    input:
        "../files/factor_percentages/{windows}/{species}.txt"
    output:
        "../data/following/factor_filtered/{windows}/{species}_done.txt"
    threads:
        99
    shell:
        "python3 ../scripts/filter_by_percentages.py {input} \
        {wildcards.species} {output}"

################################################################################
### Part A) From ranges of pure factor to FCGR
################################################################################

rule purifying:
    input:
        "../data/genomes/{species}_genomes.fna"
    output:
        "data/samples/{windows}_{n_samples}/{species}/{factors}_pure.fna"
    shell:
        "python3 scripts/masking_to_pure.py {wildcards.factors} \
            {wildcards.windows} {wildcards.n_samples} {input} {output}"

rule pure_CGR:
    input:
        "data/samples/{windows}_{n_samples}/{species}/{factors}_pure.fna"
    output:
        "data/following/CGRs/{windows}_{n_samples}/{species}/{factors}_pure_done.txt"
    threads:
        99
    shell:
        "python3 ../scripts/windowed_CGR.py {input} {wildcards.species} \
            {wildcards.windows} {wildcards.n_samples} {threads} {output}"

rule pure_FCGR:
    input:
        "data/following/CGRs/{windows}_{n_samples}/{species}/{factors}_pure_done.txt"
    output:
        "files/FCGRs/{windows}_{n_samples}_{kmer}/{species}/{factors}_pure_FCGRs.txt"
    threads:
        99
    shell:
        "python3 ../scripts/windowed_FCGR.py {input} {wildcards.species} \
            {wildcards.windows} {wildcards.kmer} {wildcards.n_samples} \
            {threads} {output}"

################################################################################
### Part B) when chosing species MDS
################################################################################

rule concatenating_each:
    output:
        "files/FCGRs/{windows}_{n_samples}_{kmer}/each_concatenated/{species}_FCGRs.txt"
    shell:
        """
        cat files/FCGRs/{wildcards.windows}_{wildcards.n_samples}_{wildcards.kmer}/{wildcards.species}/*_pure_FCGRs.txt \
            ../scaling/files/FCGRs/{wildcards.windows}_{wildcards.n_samples}_{wildcards.kmer}/{wildcards.species}_FCGRs.txt >> {output}
        """

rule FCGRs_dist_each:
    input:
        "files/FCGRs/{windows}_{n_samples}_{kmer}/each_concatenated/{species}_FCGRs.txt"
    output:
        "files/distances/manhattan/{windows}_{n_samples}_{kmer}/each_concatenated/{species}_dist_matrix.RData"
    threads:
        99
    shell:
        "Rscript ../scripts/manhattan_distance.R {input} {output}"

rule fitting_each:
    input:
        "files/distances/manhattan/{windows}_{n_samples}_{kmer}/each_concatenated/{species}_dist_matrix.RData"
    output:
        "files/distances/manhattan/{windows}_{n_samples}_{kmer}/each_concatenated/{species}_fit.RData"
    threads:
        99
    shell:
        "Rscript ../scripts/fitting_dist.R {input} {output}"

rule MDS_each:
    input:
        "files/distances/manhattan/{windows}_{n_samples}_{kmer}/each_concatenated/{species}_fit.RData"
    output:
        "files/results/{windows}_{n_samples}_{kmer}/{species}_MDS_all_factors.png"
    shell:
        "Rscript scripts/MDS_all_factors.R {output} {input} \
            {wildcards.windows} {wildcards.kmer} {wildcards.species} \
            {wildcards.n_samples}"

################################################################################
### Part C) when chosing comparison of two species into MDS
################################################################################

rule concatenating_pairwise:
    output:
        "files/FCGRs/{windows}_{n_samples}_{kmer}/pairwise_concatenated/{species}_vs_{comparison}_FCGRs.txt"
    shell:
        """
        mkdir temp
        # Copy everything to temp directory, adding the species abbreviation
        all_actual_species=$( find files/FCGRs/{wildcards.windows}_{wildcards.kmer}/{wildcards.species}/*_pure_FCGRs.txt )
        for each_actual in $all_actual_species; do
            file_name_only=$( basename $each_actual )
            sed -e 's/^/{wildcards.species}_/' $each_actual > temp/$file_name_only
        done

        all_comparison=$( find files/FCGRs/{wildcards.windows}_{wildcards.kmer}/{wildcards.comparison}/*_pure_FCGRs.txt )
        for each_comparison in $all_comparison; do
            file_name_only=$( basename $each_comparison )
            sed -e 's/^/{wildcards.comparison}_/' $each_comparison > temp/$file_name_only
        done

        # Now, can concatenate everything in one file
        cat temp/{wildcards.species}*.txt temp/{wildcards.comparison}*.txt >> {output}
        rm -r temp
        """

rule FCGRs_dist_pairwise:
    input:
        "files/FCGRs/{windows}_{n_samples}_{kmer}/pairwise_concatenated/{species}_vs_{comparison}_FCGRs.txt"
    output:
        "files/distances/manhattan/{windows}_{n_samples}_{kmer}/pairwise_concatenated/{species}_vs_{comparison}_dist_matrix.RData"
    threads:
        99
    shell:
        "Rscript ../scripts/manhattan_distance.R {input} {output}"

rule fitting_pairwise:
    input:
        "files/distances/manhattan/{windows}_{n_samples}_{kmer}/pairwise_concatenated/{species}_vs_{comparison}_dist_matrix.RData"
    output:
        "files/distances/manhattan/{windows}_{n_samples}_{kmer}/pairwise_concatenated/{species}_vs_{comparison}_fit.RData"
    threads:
        99
    shell:
        "Rscript ../scripts/fitting_dist.R {input} {output}"

rule MDS_pairwise:
    input:
        "files/distances/manhattan/{windows}_{n_samples}_{kmer}/pairwise_concatenated/{species}_vs_{comparison}_fit.RData"
    output:
        "files/results/{windows}_{n_samples}_{kmer}/{species}_MDS_pairwise.png"
    shell:
        "Rscript scripts/MDS_all_factors.R {output} {input} \
            {wildcards.windows} {wildcards.kmer}"
